api/src/test/java/io/druid/data/input/impl/PrefetchableTextFilesFirehoseFactoryTest.java

```diff
@@ -0,0 +1,388 @@
+    FileUtils.forceDelete(firehoseTempDir);
+    FileUtils.forceMkdir(firehoseTempDir);
+
+    for (int i = 0; i < 10; i++) {
+      // Each file is 1390 bytes
+      try (final Writer writer = new BufferedWriter(
+          new FileWriter(new File(testDir, "test_" + i))
+      )) {
+        for (int j = 0; j < 100; j++) {
+          final String a = (20171220 + i) + "," + i + "," + j + "\n";
+          writer.write(a);
+        }
+      }
+    }
+  }
+
+  @AfterClass
+  public static void teardown() throws IOException
+  {
+    FileUtils.forceDelete(testDir);
+    FileUtils.forceDelete(firehoseTempDir);
+  }
+
+  private static void assertResult(List<Row> rows)
+  {
+    Assert.assertEquals(1000, rows.size());
+    rows.sort((r1, r2) -> {
+      int c = r1.getTimestamp().compareTo(r2.getTimestamp());
+      if (c != 0) {
+        return c;
+      }
+      c = Integer.valueOf(r1.getDimension("a").get(0)).compareTo(Integer.valueOf(r2.getDimension("a").get(0)));
+      if (c != 0) {
+        return c;
+      }
+
+      return Integer.valueOf(r1.getDimension("b").get(0)).compareTo(Integer.valueOf(r2.getDimension("b").get(0)));
+    });
+
+    for (int i = 0; i < 10; i++) {
+      for (int j = 0; j < 100; j++) {
+        final Row row = rows.get(i * 100 + j);
+        Assert.assertEquals(new DateTime(20171220 + i), row.getTimestamp());
+        Assert.assertEquals(String.valueOf(i), row.getDimension("a").get(0));
+        Assert.assertEquals(String.valueOf(j), row.getDimension("b").get(0));
+      }
+    }
+  }
+
+  @Test
+  public void testWithoutCacheAndFetch() throws IOException
+  {
+    final TestPrefetchableTextFilesFirehoseFactory factory =
+        TestPrefetchableTextFilesFirehoseFactory.with(testDir, 0, 0);
+
+    final List<Row> rows = new ArrayList<>();
+    try (Firehose firehose = factory.connect(parser, firehoseTempDir)) {
+      while (firehose.hasMore()) {
+        rows.add(firehose.nextRow());
+      }
+    }
+
+    assertResult(rows);
+  }
+
+  @Test
+  public void testWithoutCache() throws IOException
+  {
+    final TestPrefetchableTextFilesFirehoseFactory factory =
+        TestPrefetchableTextFilesFirehoseFactory.with(testDir, 0, 2048);
+
+    final List<Row> rows = new ArrayList<>();
+    try (Firehose firehose = factory.connect(parser, firehoseTempDir)) {
+      while (firehose.hasMore()) {
+        rows.add(firehose.nextRow());
+      }
+    }
+
+    assertResult(rows);
+  }
+
+  @Test
+  public void testWithZeroFetchCapacity() throws IOException
+  {
+    final TestPrefetchableTextFilesFirehoseFactory factory =
+        TestPrefetchableTextFilesFirehoseFactory.with(testDir, 2048, 0);
+
+    final List<Row> rows = new ArrayList<>();
+    try (Firehose firehose = factory.connect(parser, firehoseTempDir)) {
+      while (firehose.hasMore()) {
+        rows.add(firehose.nextRow());
+      }
+    }
+
+    assertResult(rows);
+  }
+
```

Comment:
Can you add a test where both cache and prefetch are enabled, but the max prefetch size is less than the max cache capacity?